c/tskit/trees.c

```diff
@@ -2428,10 +2480,16 @@ get_mutation_samples(
                             &mut_samples_row, (tsk_bit_array_value_t) node);
                     }
                 }
-                mut_offset++;
             }
-            mut_offset++; // One more for the ancestral allele
-            get_allele_samples(site, &mut_samples, &out_row, &(num_alleles[site->id]));
+            site_offset += site->mutations_length + 1;
+            get_allele_samples(site, &mut_samples, &out_row, &(num_alleles[site_idx]));
+            site_idx++;
+
+            // If we've hit the last site in our index, finish up
+            if (site_idx >= n_sites) {
+                ret = 0;
+                goto out;
```

Comment:
We only use gotos for "exception handling", so it would be good to change the flow a bit here. I'm also not sure that you're correctly iterating over the trees (see comment above). How about this?

```c
while (site_idx < num_sites) {
    ret = tsk_tree_seek(&tree, sites_position[sites[site_idx]], 0);
   if (ret < 0) {
        goto out;
   }
    tsk_bug_assert(ret == TSK_TREE_OK);
    for (s = 0; s < tree.sites_length && site_idx < num_sites; s++) {
           site = &tree.sites[s];
           site_idx++;
        
          // do stuff
     } 

}
```

Reply:
Ah, I see what you're saying. After a few tests, I see what's going on. The issue I have in your proposed solution is that we'd also be calling realloc for every site. From what I understand, realloc is not specified in the C standard (GNU returns the same memory when realloc is called with the same size, but I'm not sure about other libc implementations -- see [here](https://www.gnu.org/software/libc/manual/html_node/Changing-Block-Size.html) in the portability notes).

I've proposed a solution (my most recent commit) where we seek to the first tree, then keep seeking until we run out of sites or trees.

What do you think about this? I can implement what you've proposed if you think that's the better, more straightforward route.

Reply:
Generally, things have to be correct *first* and then fast afterwards.  You make them correct by doing things simply first. Then you profile, and if it's not fast enough, you optimise **based on this profiling**. 

The realloc thing you identify is a small detail (which is trivially worked around, if it really *is* a problem). You haven't profiled this, so you don't know if it is a problem. That is  not sufficient reason to make the code so much more complicated.

I do understand the impulse to optimise, but you're going about it the wrong way. I've found this classic [worse is better](https://dreamsongs.com/RiseOfWorseIsBetter.html) essay very illuminating, and has convinced me of the importance of simplicity of implementation, almost above all other considerations.

Reply:
Thank you for passing that along, I've heard of "worse is better" colloquially, but I haven't seen this essay before. I agree with it and like the ideas, I just need to translate them into my software design.

I've adopted your suggestions and things look a bit simpler. `tsk_tree_seek` always returns 0 unless there's an error, so there's no need to make any further asserts as far as I can tell.

I also updated the python code to reflect the change.